awx/main/scheduler/task_manager.py

```diff
@@ -510,13 +487,49 @@ def process_pending_tasks(self, pending_tasks):
                 self.start_task(task, None, task.get_jobs_fail_chain(), None)
                 continue
 
+            # Determine if ther is control capacity for the task
+            remaining_control_capacity = self.get_remaining_capacity('controlplane', capacity_type='control')
+            if remaining_control_capacity < settings.AWX_CONTROL_PLANE_TASK_IMPACT:
+                logger.debug(f"Skipping task {task.log_format} in pending, not enough capacity left on controlplane to control new tasks")
+                continue
+
+            control_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(
+                task, self.graph['controlplane']['instances']
+            ) or InstanceGroup.find_largest_idle_instance(self.graph['controlplane']['instances'], capacity_type='control')
+            if not control_instance:
+                logger.debug(f"Skipping task {task.log_format} in pending, not enough capacity left on controlplane to control new tasks")
+                continue
+
+            task.controller_node = control_instance.hostname
```

Comment:
I have some concern about assigning here. I thought maybe a local variable for `controller_node` would be best, and then assign right before `start_task` is called. You could still do this, and it's not necessarily a bad idea, but it may force some surprising follow-on changes because I think it could get saved when _execution_ capacity falls short.

Reply:
Next comment, if you get to this line and `task.capacity_type == 'control'`, then you should assign `execution_node` to the same value as `controller_node`, put it in the `controlplane` IG, start task, and then `continue`. Project updates and system jobs should not search over preferred instance groups.

Reply:
did the second thing already. I will look at setting `task.controller_node` later

I _do_ think that is a bit paranoid though, because we dont call `task.save()` until the bottom of `start_task` ....which we never get to if we don't have execution capacity

Reply:
What I was thinking of was at the end if this method we do `UnifiedJob.objects.bulk_update(tasks_to_update_job_explanation, ['job_explanation'])`, but now that I paste that I realize it won't update other fields, lol.